Serverless Console Dev Mode

Danwakeem opened the pull request:
## Description
Adding a `--dev` flag to the root `serverless` command to launch serverless console dev mode activity streaming in the CLI.

## Usage

In an existing project
```bash
sls --devModeFeed # This will start the dev activity feed focused on the functions in your serverless.yml file
```

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+const apiRequest = require('@serverless/utils/api-request');
+
+const stage = process.env.SERVERLESS_PLATFORM_STAGE || 'prod';
+const socketUrl =
+  stage === 'prod'
+    ? 'wss://ehst4ktjyi.execute-api.us-east-1.amazonaws.com/prod'
+    : 'wss://4c684ym73e.execute-api.us-east-1.amazonaws.com/dev';
```
Should I move this into `@serverless/utils`? I wasn't sure if it made more sense there or just in here 🤔

Danwakeem replied:
Yes @medikoo I think either way we need to fix the urls. I am planning on updating these urls today 👍

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+
+const capitalizeFirstLetter = (string) => string.charAt(0).toUpperCase() + string.slice(1);
+
+const formatDuration = (milliseconds) => {
+  const seconds = milliseconds / 1000;
```
For all of this "UI" related stuff I am thinking I should move this out into a package that we can share between console and the CLI, possibly even set up so we can edit these configs remotely but I need to do a bit more research before I can move on this idea 👍 

So for the most part, just ignore these "UI" related utilities for now. I just didn't want this small detail to block reviews and stuff.

Danwakeem replied:
I am going to move this to `@serverless/utils` 👍

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+    streamBuff.on('data', handleSocketMessage);
+    const ws = await connectToWebSocket(context.targetFunctions, context.org);
+    // Only open for 5 minutes
+    setTimeout(() => {
```
Ahh also sorry I just realized that I need to add the still working logic. This will not stay fixed as a 5 minute limit.

Danwakeem replied:
@medikoo Okay I added some logic here to do a few things.

- Reconnect to our websocket every hour. (There is a 2 hour limit on connection in API Gateway so I figured I would reconnect well before this time limit ends)
- After 1.5 hours of inactivity pause the websocket and prompt the user to confirm they are still working. (This is so if someone leaves their computer open we won't just keep their websocket connection alive forever)

[The logic I have right now](https://github.com/serverless/serverless/pull/11782/commits/8549bf16ab60e3aa470449004e792a6b5ed982f0) is a bit confusing so if you have suggestions for improvements please let me know 🙏

medikoo commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+  let finalName = `AWS SDK • ${capitalizeFirstLetter(service)} • ${operation.toUpperCase()}`;
+  if (name.includes('aws.sdk.dynamodb')) {
+    finalName = `AWS SDK • DynamoDB • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.eventbridge')) {
+    finalName = `AWS SDK • Event Bridge • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.secretsmanager')) {
+    finalName = `AWS SDK • Secrets Manager • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.kinesis')) {
+    finalName = `AWS SDK • Kinesis • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.elastictranscoder')) {
+    finalName = `AWS SDK • Elastic Transcoder • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.iotdata')) {
+    finalName = `AWS SDK • IOT Data • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.kinesisvideo')) {
+    finalName = `AWS SDK • Kinesis Video • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.sns')) {
+    finalName = `AWS SDK • SNS • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.sqs')) {
+    finalName = `AWS SDK • SQS • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.ssm')) {
+    finalName = `AWS SDK • SSM • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.s3')) {
+    finalName = `AWS SDK • S3 • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.elb')) {
+    finalName = `AWS SDK • ELB • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.emr')) {
+    finalName = `AWS SDK • EMR • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.es')) {
+    finalName = `AWS SDK • ES • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.fms')) {
+    finalName = `AWS SDK • FMS • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.ecs')) {
+    finalName = `AWS SDK • ECS • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.ec2')) {
+    finalName = `AWS SDK • EC2 • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.eks')) {
+    finalName = `AWS SDK • EKS • ${operation.toUpperCase()}`;
+  } else if (name.includes('aws.sdk.ebs')) {
+    finalName = `AWS SDK • EBS • ${operation.toUpperCase()}`;
+  }
```
Is this a fixed list? What if we add new service, will we remember to update this module?

Maybe let's come up with one generic method and iterate just on those which require specific translation

Danwakeem replied:
I think the trouble here is that the names are kind of curated. 

e.g dynamodb = DynamoDB, eventbridge = Event Bridge, s3 = S3. 

There is a fallback if a service doesn’t exist in this curated list so that it will just use the uppercase version of the service in the `aws.sdk.service`.

I could probably just remove some of the unnecessary formats here but I am not sure this is a way to not do some sort of curation here. 🤔

medikoo replied:
Ok, yes, so in that case, let's just remove those cases which are well addressed by the fallback

medikoo reviewed (changes requested):
Thanks @Danwakeem, that looks really good 👍

Danwakeem commented on lib/cli/interactive-setup/index.js:
```diff
@@ -10,11 +10,13 @@ const steps = {
   awsCredentials: require('./aws-credentials'),
   deploy: require('./deploy'),
+  consoleResolveOrg: require('./console-resolve-org'),
+  consoleSetupIamRole: require('./console-setup-iam-role'),
```
I can double check if the old order was fine but I was seeing an issue when setting up a new integration, deploying the service, then configuring dev mode.

It took the integration service a minute to register up the new service with the integration. 

Moving the deploy step ahead of the integration set up meant that the service was already apart of the customers AWS account when the integration was being configured. This eliminated that delay so I could immediately look up the service functions in the integration inventory.

medikoo replied:
Deploy is done only for newly created services (where it's mandatory to deploy so functions are accessible), it won't be issued if command was run in context of existing service.

Danwakeem replied:
Yeah the creating a new service is the context I am referring to.

For example if you run `sls --console --enableDev` with the intent to create a new service, if we wait to deploy the service after the integration is set up we will need to add logic to ensure the new service's functions are added to the customers integration inventory before we can `enableDev` since there is usually a delay between the deployment and integration registering the new functions in the customers account with our inventory service.

Danwakeem replied:
Yeah if this will cause some reporting issues or isn't something we want to tackle right now we should probably just not implement the onboarding changes for now since it isn't a requirement for this initial MVP anyway.

I can remove the onboarding flow stuff for now 👍

medikoo replied:
I think this order should definitely remain untouched. Also it's still not clear to me why it was changed, knowing that _deploy_ runs only for new services, which cannot be integrated upfront for dev mode anyway

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+    // Add single function name or all function names to the list
+    const functionNames = [];
+    const osFunctions = [];
+    if (context.options.function) {
```
Ahh yes I need to add this as an option. I forgot to do that when I refactored this.

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+const successMessage = 'Your functions are instrumented for development.';
+const progressKey = 'instrumentation-progress';
+
+const allFunctionsExist = async (context) => {
```
This checks that all of the functions exist in the integration.

medikoo replied:
What does it mean that function exists in the integration? It means that function is integrated, or that it's visible and can be integrated?

medikoo replied:
My understanding of "integrated" is that function has already tracing enabled

Danwakeem replied:
Yes sorry, when I am saying a function existing in the integration I mean that the function has been deployed and has been registered in the inventory service. 

If you create a new function and forget to deploy that function we won’t be able to instrument it for dev mode.

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+      return false;
+    }
+
+    const { isInstrumented } = await checkInstrumentationStatus(context);
```
Yeah it would fail. When we wait for all functions to be instrumented we need be sure that all of the functions we requested are attached to our integration. If they are not then the waiting for instrumentation will stall indefinitely.

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+
+    const { isInstrumented } = await checkInstrumentationStatus(context);
+    if (isInstrumented) {
+      log.notice.success(successMessage);
```
I was copying this behavior for our set up IAM role step. In that step if we set up the integration or not we post the same success message. 

I could change both steps to what you proposed though 🤷‍♂️ personally I don’t have a preference either way.

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+    const instrumentationProgress = progress.get(progressKey);
+    instrumentationProgress.notice('Instrumenting functions', 'This may take a few minutes...');
+    // Chunk targetInstrumentations into 50 resources per request
+    let chunkedResources = [context.targetInstrumentations];
+    if (context.targetInstrumentations.length > 50) {
+      chunkedResources = context.targetFunctions.reduce(
+        (acc, functionName) => {
+          let currentIndex = acc.length - 1;
+          if (acc[currentIndex].length === 50) {
+            acc.push([]);
+            currentIndex += 1;
+          }
+          acc[currentIndex].push(functionName);
+          return acc;
+        },
+        [[]]
+      );
+    }
```
Yeah no problem 😆 I think that is a good idea given everyone hates the `reduce` function 😆

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+  chunkSize: 2048 * 1000000,
+});
+
+const getDuration = (startTime, endTime) => {
```
I believe it is the epoch time 👍 I can add a comment

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+  let customTags = {};
+  try {
+    const parsedTags =
+      typeof activity.customTags === 'object'
+        ? activity.customTags
+        : JSON.parse(activity.customTags);
```
Now that you mention it, I actually don’t know why this is like this 🤔 I will look into this.

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+    } else {
+      const names = Object.keys(context.serverless.service.functions);
+      for (const name of names) {
+        const functionName = `${context.serverless.service.service}-${context.serverless.service.provider.stage}-${name}`;
```
I get this when I try and access that function 🤔 

<img width="1119" alt="Screenshot 2023-03-09 at 8 38 46 AM" src="https://user-images.githubusercontent.com/1753824/224058324-9dcc2e01-f5c8-496c-a41c-9daddd3d587f.png">

medikoo replied:
Yes, sorry I made a typo, as apparently the function is weirdly named as `getAllFunctionsNames` -> https://github.com/serverless/serverless/blob/ad8c24ecfca4ec9dea87d9114ad3f2afcab85ae2/lib/classes/service.js#L313

Danwakeem replied:
Huh 🤔 For some reason when I run this in the context of a service directory I getting undefined for all of the function names 👇 

<img width="961" alt="Screenshot 2023-03-09 at 9 17 28 AM" src="https://user-images.githubusercontent.com/1753824/224069091-ee6833cc-af3b-4359-aeef-83220ba06150.png">

medikoo replied:
Indeed, I've checked and for that to work first `service.setFunctionNames(options)` needs to be called, and that one is normally run in context of `serverless.run()` which would only happen if we're after _deploy_.

I guess it's best then to call this method before calling `getAllFunctionsNames`

Danwakeem replied:
This works great 😎 Thanks man!

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+      const names = Object.keys(context.serverless.service.functions);
+      for (const name of names) {
+        const functionName = `${context.serverless.service.service}-${context.serverless.service.provider.stage}-${name}`;
+        // console.log(functionName);
```
🤦

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+
+    const { allExist, total, functionCount } = await allFunctionsExist(context);
+    if (!allExist) {
+      log.notice(
+        `Only ${functionCount} of ${total} functions exist on this integration. Be sure to deploy the service before using dev.`
+      );
```
Possibly 🤔 I was thinking of preventing dev mode from booting up because if there are local changes you would need to deploy those changes before you could see those changes reflected in dev mode anyway.

That being said, it would probably be worth it if we just switch this to a warning. I can work on this today.

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+'use strict';
```
Yeah I agree. I like that first name suggestion better 😎

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode.js:
```diff
@@ -0,0 +1,337 @@
+
+const capitalizeFirstLetter = (string) => string.charAt(0).toUpperCase() + string.slice(1);
+
+const formatDuration = (milliseconds) => {
```
Yeah sure 👍 I can use this package.

Danwakeem commented on lib/cli/interactive-setup/deploy.js:
```diff
@@ -34,6 +34,11 @@ module.exports = {
       return false;
     }
 
+    if (devCommand) {
+      context.inapplicabilityReasonCode = 'CONSOLE_DEV_COMMAND';
+      return false;
+    }
+
```
Ahh okay you are right, however, I think I need to change the previous if statement from looking at `options.console` to `isConsole` so I can remove this 👍

Danwakeem commented on lib/cli/interactive-setup/deploy.js:
```diff
@@ -91,6 +95,7 @@ module.exports = {
     context.serverless = serverless;
 
     printMessage();
+    progress.clear();
```
Since I moved the deploy step above the console steps if I didn't clear the progress here there was an `Updating` progress that would persist through the rest of the prompts. I was assuming this was just not seen before because this was the final step 👇 

<img width="1296" alt="Screenshot 2023-03-09 at 9 14 59 AM" src="https://user-images.githubusercontent.com/1753824/224068518-572469db-270c-4a5f-97c6-33ba273166c7.png">

medikoo replied:
Indeed, internal progress is not cleared as it's assumed that it'll be cleared at end of processing in general script logic. So clearing it here makes sense

Danwakeem commented on lib/cli/interactive-setup/deploy.js:
```diff
@@ -53,8 +58,7 @@ module.exports = {
     // We want to proceed if local credentials are available
-    if (new AWS.Config().credentials) return true;
-
+    if (new AWS.Config().credentials || context.awsAccountId) return true;
```
For some reason the deploy step was actually getting skipped for me when I removed the `context.awsAccountId`. It is claiming that `new AWS.Config().credentials` is `null` 👇 

<img width="1293" alt="Screenshot 2023-03-09 at 9 20 42 AM" src="https://user-images.githubusercontent.com/1753824/224069926-703a1b28-f51d-486a-a9c0-17f601131107.png">

medikoo replied:
Sure, still it's nicer to fix it, in a way I've proposed above

Danwakeem commented on lib/cli/interactive-setup/console-instrument.js:
```diff
@@ -0,0 +1,202 @@
+      return false;
+    }
+
+    if (!enableDev) {
```
Since we are removing the onboarding steps there is no need to prompt.

medikoo replied:
> Since we are removing the onboarding steps 

What steps are we removing?

Danwakeem replied:
Sorry I should have specified, removing the **_dev mode_** onboarding steps 😅

medikoo replied:
This PR is about adding dev mode onboarding steps. So you mean it's not longer applicable?

Danwakeem replied:
The task I was assigned in linear was originally just supposed to be activating the dev mode feed on an existing serverless project.

There really wasn't a requirement for us to add the enable dev mode or dev mode feed to the CLI onboarding steps. I just thought it would be simple enough to add dev mode to the CLI "onboarding" and existing projects since they are the same general steps.

But if we can't move the `deploy` step around in the onboarding flow then we can't allow people to enable dev mode when creating a new project since the deploy happens after the integration is created and the delay can cause the dev mode instrumentation to fail.

medikoo replied:
In my understanding, what this task needs is in 80% covered by onboarding, and it simply feels as an onboarding extension.

> But if we can't move the deploy step around 

Still, it's not clear why there's a need for that. In the case of (1) existing service, `deploy` will not happen, and in case of (2) not existing service, we do not apply dev mode anyway. So to me, it feels like `deploy` is totally irrelevant.

Danwakeem replied:
You are correct and I think we are saying the same thing here 😅 I reverted the steps in the index file so I think we should be good or did I need to move one of the new steps around?

medikoo replied:
ok, I got lost a bit 😅

medikoo commented on lib/cli/interactive-setup/index.js:
```diff
@@ -12,6 +12,8 @@ const steps = {
   dashboardLogin: require('./dashboard-login'),
   consoleResolveOrg: require('./console-resolve-org'),
   consoleSetupIamRole: require('./console-setup-iam-role'),
+  consoleEnableDevMode: require('./console-enable-dev-mode'),
+  consoleDevModeFeed: require('./console-dev-mode-feed'),
   dashboardSetOrg: require('./dashboard-set-org'),
   awsCredentials: require('./aws-credentials'),
   deploy: require('./deploy'),
```
I think best if those new steps are last. Logically it feels more right, and we change a bit rules by which they're applied it should remain working

Danwakeem commented:
Going to wait till I cut a new release for utils and make the changes to this branch before requesting a re-review 😎

Danwakeem commented:
Okay got the utils stuff updated but I am going to wait for confirmation from Austen and the gang on the command name before requesting re-review 🙏

Danwakeem commented:
Okay I finally requested the re-review on this one 🙏

- I moved the command to a root `dev` command
- Updated the utils library to the new version with the console ui functions
- Added a runtime compatibility check so that we reject projects that are not compatible with dev mode

Danwakeem commented on lib/cli/interactive-setup/deploy.js:
```diff
@@ -1,7 +1,7 @@
 'use strict';
 
 const Serverless = require('../../serverless');
-const { writeText, style, log } = require('@serverless/utils/log');
```
I can revert the changes in this file if we would like but I feel like the changes here might be worth making 🤷

medikoo replied:
@Danwakeem they're good, but best if either they're proposed in different PR (as they're not necessary related), or we can keep it in one PR, assuming we stick to semantic commits and PR can be rebased instead of squashing

Danwakeem replied:
Okay yeah this is a good call. I Can open a separate PR 👍

medikoo reviewed (commented):
@Danwakeem Instead of configuring it as totally separate command and copy paste some of internal logic of interactive CLI, wouldn't it be better to make `serverless dev` work as an alias to `serverless --dev` ?

I have a notion that this would be cleaner (no code duplication), and require less implementation effort

medikoo commented:
@Danwakeem to further clarify what I have in mind. I think cleanest implementation is as follows:
- Add to schema `--dev` option for interactive CLI command
- Configure `consoleEnableDevMode` and `consoleDevModeFeed` steps, which are not applicable if `!options.dev`
- Ensure in commands handling that `serverless dev` is translated to `serverless --dev`
- (optional) in interactive CLI crash with meaningful message if `options.dev` and we're not in service context

Danwakeem commented:
Thanks for the clear breakdown @medikoo 👍 as usual I do like this idea. That way I don’t have to re-create the index file. I can make this change this morning.

medikoo commented:
@Danwakeem concerning mapping `serverless dev` into `serverless --dev`. I think we should follow with a similar method with which we map  `sls deploy -f <function>` into `sls deploy function -f <function>` -> https://github.com/serverless/serverless/blob/16dd286a1d8aa764003565bcd6d7334be290e9bf/scripts/serverless.js#L137-L149

I guess the best is to refactor this part so it handles both aliases

Danwakeem commented:
@medikoo that feels more confusing that simply adding a dev file to the plugins list. Is there any particular reason we should go that route rather than what I had put in my most recent commit? 🤔

medikoo commented:
> Is there any particular reason we should go that route rather than what I had put in my most recent commit? 🤔

`serverless dev` is expected to work exactly as onboarding (so `serverless` command) serving a specific case (service to be setup for console + dev mode enabled).

Maintainancence wise it feels more future-proof, to have _onboarding_ set up in a single place, and just tweak how it works depending on the options. Instead of configuring two different onboarding commands with two very different approaches to how commands are set up and have code duplicated among them.

medikoo commented:
Additionally relying on onboarding steps in a command configured in context of an instance, introduces precedence that can turn quite problematic in the future.

We already call an instance `deploy` command in the context of the onboarding. Now if we will have the request to call `dev` command (which on its own runs same onboarding steps) because e.g. we will now like to have same behavior when users are setting up new service, it can get really quirky and buggy

Safest is to have clean separation between onboarding and instance logic, so no unwanted processing flow mix can happen. We will also end with smaller PR that way, so definitely less technical debt introduced.

Danwakeem commented:
@medikoo Okay I just updated the implementation to not include the custom plugin. Now that I understand that block you were pointing me to I see why this is a bit simpler now 😅 

Let me know if there is any other tweak I should make to this 🙏

medikoo commented on scripts/serverless.js:
```diff
@@ -140,6 +140,12 @@ processSpanPromise = (async () => {
       const commands = args.slice(0, firstParamIndex === -1 ? Infinity : firstParamIndex);
+
+      if (commands.join('') === 'dev') {
+        process.argv[2] = '--dev';
```
Let's also update the comment above, that it's also about `serverless dev` to `serverless --dev`

medikoo reviewed (commented):
Thanks @Danwakeem, looks very good 👍 I've just proposed some organization improvements which I believe should further simplify this PR

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode-feed.js:
```diff
@@ -0,0 +1,218 @@
+          consoleMonitoringCounter.logBatches += 1;
+          process.stdout.write(chalk.grey(`${time} • ${resourceName} • Log\n`));
+          try {
+            const parsedBody = JSON.parse(activity.body);
+            if (typeof parsedBody === 'string') {
+              throw new Error('Not a JSON object');
+            }
```
Yes. In the events that a customer sends a string response from their lambda function, for example `"Foo bar"`, this will parse correctly in the `JSON.parse` function and be returned correctly from the customers lambda function but will not be a JSON object 👍

medikoo replied:
Ok, so wouldn't be more natural to write following:

```javascript
try {
  if (noObject) throw;
  // object case
} case (error)
 // string case
}
```

as:

```javascript
if (noObject) // string case
else // object case
```

So without try/catch handling?

Danwakeem replied:
This would not work. We need the try/catch because we want to pretty print any logs that are formatted as json but logs are not always JSON objects so if the log message was a simple string or a non json object this would throw an error.

Danwakeem commented on lib/cli/interactive-setup/console-dev-mode-feed.js:
```diff
@@ -0,0 +1,218 @@
+        return arr;
+      }
+    }, []);
+    if (!Array.isArray(jsonArray)) return;
```
Updated to avoid `reduce` 😎 Note we need to check the array because when we first establish a connection with filters applied or not the websocket sends a successful connection message that is not an array.

Danwakeem replied:
Okay never mind I couldn't get the flat map thing to work in all scenarios for some reason where as the reduce is working 🤷

Danwakeem replied:
Ignore my deleted comment 😆 It wasn't the flat map thing.

medikoo commented on lib/cli/commands-schema/no-service.js:
```diff
@@ -30,6 +30,16 @@ commands.set('', {
+    },
+    'function': {
+      usage:
+        'Launch dev mode activity feed focused on a specific function inf your serverless.yml file.',
```
I think just this argument is not responsible for launching dev mode activity.

More appropriate description will probably be something as: _Name of a function to be observed in dev mode feed_

medikoo commented on lib/cli/interactive-setup/console-dev-mode-feed.js:
```diff
@@ -0,0 +1,187 @@
+  try {
+    const splitData = data.toString('utf-8').split(';;;');
+    const jsonArray = splitData
+      .filter((item) => item !== '' && Array.isArray(JSON.parse(item)))
```
Wouldn't `item.startsWith('[') be good enough here? (relying on `JSON.parse` in filter and then repeating it in map doesn't feel optimal)

Danwakeem replied:
This is a good idea, thanks for the suggestion 😎

medikoo commented on lib/cli/interactive-setup/index.js:
```diff
@@ -63,6 +66,21 @@ module.exports = async (context) => {
+    if (!devModeRuntimeCompatibility.includes(provider.runtime)) {
+      context.incompatibleRuntime = true;
+      log.error('This services runtime is not currently supported by Serverless Console Dev Mode.');
+      throw new Error('INCOMPATIBLE_RUNTIME');
```
Do we really want to throw here?

Even if, we should use `ServerlessError` and configure it with human friendly message which will be displayed to user

Danwakeem replied:
This is a good question. It wasn’t clear to me if there was a way for me to end the execution early without throwing. If I put a return here it threw a lint error 🤔

medikoo replied:
Question, is do we want here (1) abort the command and present user with the error message then we should throw (but as mentioned above using dedicated interface and with human friendly message)
Or do we want to  (2) just show error stating that given step cannot be accomplish and let command follow with other steps (as e.g. consol integration). In this case we should just log the error, do not set `context.isConsoleDevMode: true`  and let process continue

Danwakeem replied:
🤔 I guess it might be fine to just let the process continue and just log the error. That would probably be the least inconvenient.

medikoo reviewed (commented):
Thanks @Danwakeem ! Few last comments..

medikoo reviewed (commented):
@Danwakeem last minor nit pick, and we're good to go 👍

Danwakeem commented on lib/cli/interactive-setup/index.js:
```diff
@@ -63,6 +66,20 @@ module.exports = async (context) => {
+    const devModeRuntimeCompatibility = compatibilityMap.mode.dev.runtimes;
+    const { provider } = context.serverless.service;
+    if (!devModeRuntimeCompatibility.includes(provider.runtime)) {
+      context.incompatibleRuntime = true;
```
Ahh yes good point 👍

medikoo reviewed (approved):
Great work @Danwakeem 👍

Danwakeem commented:
Thanks for all your help @medikoo 🙏
